cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java

```diff
@@ -196,38 +218,71 @@ public synchronized RuntimeInfo lookup(final RunId runId) {
       }
 
       for (TwillController controller : liveInfo.getControllers()) {
-        RunId runId = controller.getRunId();
-        if (result.containsKey(runId)) {
+        RunId twillRunId = controller.getRunId();
+        if (isTwillRunIdCached(twillRunId)) {
           continue;
         }
 
         Id.Program programId = Id.Program.from(matcher.group(2), matcher.group(3), type, matcher.group(4));
-        RuntimeInfo runtimeInfo = createRuntimeInfo(type, programId, controller);
+        twillProgramInfo.put(programId, twillRunId, controller);
+      }
+    }
+
+    if (twillProgramInfo.isEmpty()) {
+      return ImmutableMap.copyOf(result);
+    }
+
+    List<RunRecord> activeRunRecords = store.getRuns(ProgramRunStatus.RUNNING, 100);
```

Comment:
With that list of `RunRecord`, you can just loop over the run record list (which shouldn't be too big, and should be of the same size as the `twillProgramInfos` table) and constructs RuntimeInfo by performing lookup into the `twillProgramInfos` table.